spackbot/handlers/style.py

```diff
@@ -23,3 +31,99 @@ async def style_comment(event, gh):
                 number,
             )
             await gh.post(comments_url, {}, data={"body": comments.style_message})
+
+
+def is_up_to_date(output):
+    """
+    A commit can fail if there are no changes!
+    """
+    return "branch is up to date" in output
+
+
+async def fix_style(event, gh):
+    """
+    Respond to a request to fix style.
+    We first retrieve metadata about the pull request. If the request comes
+    from anyone with write access to the repository, we commit, and we commit
+    under the identity of the original person that opened the PR.
+    """
+    response = requests.get(event.data["issue"]["pull_request"]["url"])
+    pr = response.json()
+
+    # Get the sender of the PR - do they have write?
+    sender = event.data["sender"]["login"]
+    repository = event.data["repository"]
+    collaborators_url = repository["collaborators_url"]
+
+    # If they don't have write, we don't allow the command
+    if not await helpers.found(gh.getitem(collaborators_url, {"collaborator": sender})):
+        logger.info(f"Not found: {sender}")
+        return (
+            "Sorry %s, I cannot do that for you. Only users with write can make this request!"
+            % sender
+        )
+
+    # Tell the user the style fix is going to take a minute or two
+    message = "Let me see if I can fix that for you! This might take a moment..."
+    await gh.post(event.data["issue"]["comments_url"], {}, data={"body": message})
+
+    # Get the username of the original committer
+    user = pr["user"]["login"]
+
+    # We need the user id if the user is before July 18. 2017
+    email = helpers.get_user_email(user)
+
+    # We need to use the git url with ssh
+    branch = pr["head"]["ref"]
+    full_name = pr["head"]["repo"]["full_name"]
+    clone_url = "git@github.com:%s.git" % full_name
+
+    # At this point, we can clone the repository and make the change
+    with helpers.temp_dir() as cwd:
+
+        # Clone a fresh spack develop to use for spack style
+        git("clone", "--depth", "1", "https://github.com/spack/spack", "spack-develop")
+
+        spack = sh.Command(f"{cwd}/spack-develop/bin/spack")
+
+        git("clone", "-b", branch, clone_url)
+        os.chdir("spack")
+        git("config", "user.name", user)
+        git("config", "user.email", email)
+
+        # This will authenticate the push with the added ssh credentials
+        git("remote", "set-url", "origin", clone_url)
+
+        # We need to add develop remote for this to work
+        git("remote", "add", "upstream", "https://github.com/spack/spack")
+        git("fetch", "upstream")
+
+        # This won't work if we are already on a develop branch
+        if branch != "develop":
+
+            git("checkout", "--track", "upstream/develop")
+            git("checkout", branch)
```

Comment:
Instead of doing another deep clone and another fetch of `develop` here, make a local copy of the `spack-develop` you already cloned, then fetch the new branch (it's a lot faster because you already have the full history locally):

Instead of doing another deep clone here, make a local copy of `spack-develop` and just fetch the new branch (it's a lot faster because you already have the full history locally):

```suggestion
        # existing clone already has `develop` fetched and checked out
        shutil.copy("spack-develop", "spack")
        os.chdir("spack")
        git("config", "user.name", user)
        git("config", "user.email", email)

        # This will authenticate the push with the added ssh credentials
        git("remote", "rename", "origin", "upstream")
        
        # probably want to rename clone_url to fork_url or something
        # since we don't clone it anymore
        git("remote", "add", "origin", clone_url)

        # we're on upstream/develop. Fetch and check out just the PR branch
        git("fetch", "origin", f"{branch}:{branch}")
        git("checkout", branch)
```

Note that this probably seems more natural for the last two lines:

```python
        git("fetch", "origin")
        git("checkout", branch)
```

But I think that won't work in the case that the branch name exists *both* in `upstream` and `origin` -- it'll error ou and say it's ambiguous.  Doing the explicit fetch gets you just that branch and it creates the local branch so that the checkout is unambiguous.

Reply:
I don't know why, but this doesn't work - it freezes on the clone and spackbot stops responding.

Reply:
What part doesn't work?  Cloning via `ssh`?  Confused because there is no clone in my modified version -- the initial clone is the same.

Reply:
Did you mean that this clone doesn't work? https://github.com/spack/spack-bot/pull/22#discussion_r678384010

Reply:
I don't know what the issue was, but I'd trigger the bot in the PR and I would see the clone command in the logs and then it would just freeze like that. The ssh change in another comment worked fine, but this particular chunk did not.

Reply:
Can you debug it and see where it is actually hanging?

Reply:
Hung here for like 5 mins

![image](https://user-images.githubusercontent.com/814322/127376798-f2e0a7ba-c7fe-40cb-bf98-027ce51909c6.png)

And previously I had killed it but I let it run this time, and I arrived at this (and note I had to change shutil.copy to shutil.copytree because copy is for a file, not directory, and now we have some recursive thing going on:

![image](https://user-images.githubusercontent.com/814322/127377171-6c771eda-0822-4c43-b5a6-d7ae08fec8a3.png)

By the way - this is just the top of the screen - the output is like 10x longer and then it finally issues a 500 internal server error. So probably shutil copytree of the repository is a bad idea? lol

Reply:
see "too many levels of symbolic links." We would need to do some kind of custom copy strategy which sounds overly complicated imho.

My thinking is that it works well as is currently implemented, this seems a bit buggy, so we should keep / try out as is until there is evidence that it's not good enough or needs to be changed.

Reply:
The bug is really simple.  Change this:

```python
shutil.copy("spack-develop", "spack")
```

to this:

```python
shutil.copy("spack-develop", "spack", follow_symlinks=False)
```

Reply:
There is a symlink in `lib/spack/docs` that points up to the top-level spack directory, so it's just recursive.  I don't know why `shutil` follows symlinks by default -- that seems like a bad default... but we've run into that before.  Sigh.

Reply:
copy is for files, I think we want:

```python
# existing clone already has `develop` fetched and checked out
shutil.copytree("spack-develop", "spack", symlinks=False)
```

Reply:
Reworked fix to use `copytree` as you pointed out.

Reply:
It still takes a lot longer than the other way, and now I'm seeing a new error that I don't know what is.
```bash
ot_1  | INFO:spackbot:Received event <gidgethub.sansio.Event object at 0x7ff4f40cf650>
spackbot_1  | INFO:sh.command:<Command '/usr/bin/git clone --depth 1 git@github.com:spack/spack spack-develop', pid 17>: process started
smee_1      | Event { type: 'error', message: undefined }
smee_1      | Connected https://smee.io/xxxxxxxxxx
```
I still think we should keep the current way.

Reply:
Ok anyway we resolved this offline and got rid of shutil (python copying is slow) and eliminated one remote clone.  yay!

Reply:
yes yay! TLDR: shutil is just really terrible.